### ISSUE TITLE
Compare against… dialog only shows after you press an additional key

### ISSUE THREAD

**Issue by SCdF** [jisaacks/GitGutter#446]
## Description

When I change what I'm comparing against, once I select the option in the command palette (eg compare against branch) and the command palette disappears, I have to press another button before the list of available options (ie branches) appears. This happens on all of these options: branches, commits, tags.

## Support Info

- Sublime Text 3142
- Platform: osx
- Arch: x64
- GitGutter 1.7.4
- Install via PC: True
- git version 2.14.1
- mdpopups 2.1.1
- markdown 2.6.6
- pygments 2.1a0
- jinja2 2.8

## Steps to Reproduce Issue

Here is a video: https://imgur.com/a/D0Cyg

In the first example it doesn't show until I hit enter. In the second I hit an arrow key. Any key works.

**Comment by deathaxe** [comment]
The "Compare against ..." commands call git to read the required data asynchronously and then the `view.window().show_quick_panel()` API function is called to display the quick panel with the compare targets. This may cause a little gap between hiding the command pallet and showing up the new quick panel, but I can't find a reason why it should require another key press to show.

It works properly on Windows, at least.

If I had to guess I'd say the panel is created but ST misses the initial paint call, required to show it up. If you then press any key or maybe even click with the mouse somewhere a repaint is requested and the panel is displayed. Maybe it's related to the asynchronous task the quick panels are created in?

Is this issue related to the specific built or does it exist with earlier builts as well?

I wrote another package "Skins" which calls the `show_quick_panal()` asynchronously. Could you check, whether it behaves the same way?

Maybe we need to ask @wbond to provide some hints. Otherwise I'd need to make the calls synchronous, which might cause the GUI to freeze on large repos or slow network drives. I would like to avoid that.

**Comment by SCdF** [comment]
> If you then press any key or maybe even click with the mouse somewhere a repaint is requested

Right, clicking with the mouse works too, specifically inside the editor frame. If you click a different editor frame it will open up the quick panel in that frame.

> Is this issue related to the specific built or does it exist with earlier builts as well?

It's been around for awhile honestly. I could install it via git instead of PC and revert tags until it works properly. I can do that in a bit.

> I wrote another package "Skins" which calls the show_quick_panal() asynchronously. Could you check, whether it behaves the same way?

Skins works fine (cool package btw).

**Comment by SCdF** [comment]
Wow OK.

So:
 - Installing it via git works for master and the most recent tag. Note that a you do this by putting the code in `~/Library/Application Support/Sublime Text 3/Packages/Gitgutter`
 - Installing it via PC does not work off the bat: it exhibits the issue. Note that at this stage *there is no code in `~/Library/Application Support/Sublime Text 3/Packages/Gitgutter`. I don't know much about ST internals, but I guess it's zipped up and held somewhere else.
 - If you use [Package Resource Viewer](https://github.com/skuroda/PackageResourceViewer) to extract the package, the code is now available in `~/Library/Application Support/Sublime Text 3/Packages
/Gitgutter`, and *the package now works correctly*.

So, if it's extracted to the `Packages` dir it works fine, if it's wherever ST packages are kept internally it doesn't work.

I'm unable to test different tags and also have the code in ST's internal location, unless you know how to install old versions through ST (or, you know where I can dump a zip or something).

**Comment by deathaxe** [comment]
Interesting but strange facts.

If you install the package via PC, it is copied to `~/Library/Application Support/Sublime Text 3/Installed Packages/Gitgutter.sublime-package` which is just the zipped GitGutter folder.

I can't imagine a reason for this issue to exist only in the zipped GitGutter version. Everything is pure OS independent python code, which should run off the zip the same way it does from the extracted folder. In deed it does no Windows. Just double checked it by installing GitGutter via PC on a second ST setup.


What happens, if you open the quick panels by key bindings (e.g. "super+shift+option+c", "c"`) or just running the command from console `view.run_command("git_gutter_compare_file_commit")`?

Does the panel show immediately or do you need any key press, too?

**Comment by SCdF** [comment]
Using the key binding or directly running the command works, yep.

So, I disabled *all* my other packages, and it works now.

However, as I'm re-enabling packages I am having trouble finding a route cause. It doesn't seem to be a specific package. In fact, I've managed to get it into a situation where branches and commits always has the bug, but commits works fine.

So I'm wondering if packages are stomping over each other somehow, and it's related to the order they get loaded in? It's very confusing. If extracted packages get initialised first that would explain why that makes it work.

**Comment by deathaxe** [comment]
Extracted package get loaded later then zipped ones.

There is no difference in handling commits/branches/... quick panels. All of them are implemented using one text command which calls git and shows the parsed output the same way. Therefore it sounds like some kind of race condition. Really strange.

**Comment by SCdF** [comment]
Right, so something is clobbering something of yours, but if it's extracted that doesn't happen because it gets loaded after other packages.

Yeah this is wonky af. I've seen it different over restarts with the same packages. I've seen it fail the first time, then work fine after that. I've seen it work for some commands and not others.

If you want you could try installing https://packagecontrol.io/packages/xpath . This seems to be one package that has some effect. However, enabling just PC, gitgutter and xpath works by itself, so it's not a complete minimal case.

If this is a ST bug it's not a new one. I tried dev builds back to Feb 2015 (3071) and it's still a problem (side note: wow is config not backwards compatible…).

Perhaps it's a PC bug? 

If you're interested, here is my entire package control user settings (ie all my installed packages):

```json
{
	"bootstrapped": true,
	"in_process_packages":
	[
	],
	"installed_packages":
	[
		"ASCII Decorator",
		"BetterFindBuffer",
		"BracketHighlighter",
		"Case Conversion",
		"Clipboard Diff",
		"Comments Aware Enter",
		"DevDocs",
		"DocBlockr",
		"Dockerfile Syntax Highlighting",
		"EDN",
		"FontCycler",
		"Git Config",
		"GitGutter",
		"Hyperion for gettext",
		"Indent XML",
		"Insert Nums",
		"InsertDate",
		"Jai",
		"JsFormat",
		"LessImproved",
		"Markdown Preview",
		"MarkdownEditing",
		"Maybs Quit",
		"MoveTab",
		"NeoVintageous",
		"open-sesame",
		"Origami",
		"Package Control",
		"PackageResourceViewer",
		"PlainTasks",
		"PostgreSQL Syntax Highlighting",
		"Rust Enhanced",
		"Sass",
		"SideBarEnhancements",
		"Six",
		"SqlBeautifier",
		"sublime-github",
		"SublimeGit",
		"SublimeLinter",
		"SublimeLinter-contrib-htmllint",
		"SublimeLinter-jshint",
		"SublimeREPL",
		"Swift",
		"SyncedSideBar",
		"TabsExtra",
		"Terminal",
		"TodoReview",
		"TOML",
		"TypeScript",
		"xpath"
	],
	"repositories":
	[
	]
}
```

And here is what I have ignored:

```json
	"ignored_packages":
	[
		"DocBlockr",
		"Markdown",
		"Six",
		"TypeScript",
		"Vintage",
		"Vintageous"
	],
```

If I set my ignore packages to:

```json
	"ignored_packages":
	[
    "DocBlockr",
    "Markdown",
    "PackageResourceViewer",
    "PlainTasks",
    "PostgreSQL Syntax Highlighting",
    "Rust Enhanced",
    "Sass",
    "SideBarEnhancements",
    "Six",
    "SqlBeautifier",
    "sublime-github",
    "SublimeGit",
    "SublimeLinter",
    "SublimeLinter-contrib-htmllint",
    "SublimeLinter-jshint",
    "SublimeREPL",
    "Swift",
    "SyncedSideBar",
    "TabsExtra",
    "Terminal",
    "TodoReview",
    "TOML",
    "TypeScript",
    "Vintage",
    "Vintageous",
    "xpath"
	],
```

**Then it works consistently**. And if I enable `xpath` (by commenting it out from this list) **then it fails consistently**.

But it's not just the `xpath` package. If I combine my standard small ignore list with also ignoring `xpath` there are still problems.

Also, this is almost certainly complicated by my packages directory, because this changes the initialisation order:

```
scdf at SCdF in ~/Library/Application Support/Sublime Text 3/Packages
$ ll
total 8
drwxr-xr-x  18 scdf  staff   612  2 Feb  2017 ASCII Decorator
drwxr-xr-x  18 scdf  staff   612  1 Sep 15:00 BetterFindBuffer
drwxr-xr-x   3 scdf  staff   102 15 Sep 14:53 Default
drwxr-xr-x  13 scdf  staff   442 27 Feb  2017 DevDocs
drwxr-xr-x  12 scdf  staff   408  2 Feb  2017 PackageResourceViewer
drwxr-xr-x  44 scdf  staff  1496  7 Aug 17:27 PlainTasks
drwxr-xr-x  24 scdf  staff   816 20 Aug 21:56 Six
drwxr-xr-x  32 scdf  staff  1088 12 Apr 13:21 SublimeLinter
drwxr-xr-x  33 scdf  staff  1122  2 Feb  2017 SublimeREPL
drwxr-xr-x  25 scdf  staff   850  1 Sep 11:01 Terminal
drwxr-xr-x  17 scdf  staff   578  1 Aug 17:35 TodoReview
drwxr-xr-x  40 scdf  staff  1360 18 Apr 09:36 TypeScript
lrwxr-xr-x   1 scdf  staff    37  2 Feb  2017 User -> /Users/scdf/Dropbox/Apps/Sublime/User
drwxr-xr-x   9 scdf  staff   306 23 Mar 21:38 backrefs
drwxr-xr-x   7 scdf  staff   238  1 Aug 18:36 dateutil
drwxr-xr-x  10 scdf  staff   340 15 Sep 12:01 lxml
drwxr-xr-x   7 scdf  staff   238 12 Apr 14:11 markupsafe
drwxr-xr-x  14 scdf  staff   476  3 Aug 11:02 mdpopups
drwxr-xr-x   6 scdf  staff   204 16 Feb  2017 pygments
drwxr-xr-x   6 scdf  staff   204 16 Feb  2017 python-jinja2
drwxr-xr-x   6 scdf  staff   204 15 Sep 12:01 python-markdown
drwxr-xr-x   7 scdf  staff   238 27 Apr 13:02 pytz
drwxr-xr-x   6 scdf  staff   204 16 Feb  2017 pyyaml
drwxr-xr-x  11 scdf  staff   374  9 Jul 16:49 shellenv
drwxr-xr-x   3 scdf  staff   102  2 Jun 10:42 sublime-github
```

It's all kinds of crazy.

**Comment by SCdF** [comment]
So I reset my packages folder, it's not just dependencies and a link to my user settings.

I've gone through each package, including restarts, and I've whittled it down to this:

```json
	[
		"DocBlockr",
		"Markdown",
		"Six",
		"TypeScript",
		"Vintage",
		"Vintageous",

		"PackageResourceViewer",
		"PlainTasks",
		"Rust Enhanced",
		"SublimeGit",
		"TOML", // If enabled, sometimes works and sometimes doesn't, same session
		"xpath"
	],
```

The first half is what I always have disabled, and the second half are problematic packages.

If I invert this so that **only** packages I have enabled are Gitgutter, PC, and the 6 packages in the second half, I can reproduce the issue.

**Comment by SCdF** [comment]
I think it's time to get the package control people / sublime involved (cc. @wbond). There are lots of weird variations here.

Out of:
```
"PackageResourceViewer",
"PlainTasks",
"Rust Enhanced",
"SublimeGit",
"TOML",
"xpath"
```

If only they (and the standard stuff) is disabled then then the issue doesn't occur.

If I disabled EVERYTHING (except PC and GG) and enable them top to bottom, the issue only occurs once I hit `xpath`.

If I disabled EVERYTHING (except PC and GG) and enable them **bottom to top**, the issue only occurs once I hit `sublimegit`.

In other words, I can enable `xpath` and `toml` without causing a problem, but if I enable everything else except `xpath` it works *until* I enable `xpath`.

I'm going to head out now, but yeah I have no idea. There isn't much more I can do, though if you want me to try things I'm happy to. Sounds very much like it's a problem external to your package specifically though, and is some issue with Sublime / PC.

**Comment by wbond** [comment]
I would open your Console, `import sys` and check out your `sys.path`. My hunch is something in there, although I could be completely wrong.

**Comment by deathaxe** [comment]
I installed all of those packages, but I still can't reproduce on Windows.

**Comment by SCdF** [comment]
Hey @wbond, unfortunately my path does not change if I change my configuration between one that works and one that doesn't.

For reference, my current path is:
```
['/Applications/Sublime Text.app/Contents/MacOS', '/Applications/Sublime Text.app/Contents/MacOS/python3.3.zip', '/Users/scdf/Library/Application Support/Sublime Text 3/Lib/python3.3', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/pygments/all', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/lxml/st3_osx_x64', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/shellenv/all', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/backrefs/st3', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/dateutil/all', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/markupsafe/all', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/python-markdown/st3', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/pytz/all', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/pyyaml/st3', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/python-jinja2/all', '/Users/scdf/Library/Application Support/Sublime Text 3/Packages/mdpopups/st3']
```

This is from the "If I disabled EVERYTHING (except PC and GG) and enable them bottom to top, the issue only occurs once I hit sublimegit." use case, and doesn't change once I hit sublimegit, including between restarts.

**Comment by deathaxe** [comment]
I've created a debug branch and added a print command right before the function call which shows the quick panel. If you want, you can
1. download https://github.com/jisaacks/GitGutter/archive/debug.zip
2. rename the file to GitGutter.sublime-package
3. replace the existing one in ~/Application Support/Sublime Text 3/Installed Packages
4. open sublime text's console 
5. open the command pallet and run one of the Compare commands

I'd expect you to immediately get a message like `GitGutter: Show branches...`, with a panel being displayed or not. In this case it's really a paint issue of ST. If the console message appears together with the panel on a key press, something is going wrong in GitGutter, which we need to investigate.

**Comment by SCdF** [comment]
I would be delighted!

Unfortunately it doesn't look like it's a paint issue: the console log message does not appear straight away, and is blocked just like the panel is behind me hitting another key / clicking.

I have no idea if this matters, because it's not clear where it's coming from, but I often get this kind of things in the console:

```
top level value must be an array
indexing: crawldb not available, indexing abandoned
```

I haven't noticed it impacting anything obviously though, and it's around whether this bug presents or not, but I thought I'd mention it.

**Comment by deathaxe** [comment]
> indexing: crawldb not available, indexing abandoned

This is most likely caused by a corrupted index database. Should be fixed by clearing `~/Application Support/Sublime Text 3/Index` folder. The index is rebuilt upon next start. This sometimes happens.

> top level value must be an array

In rare cases I also saw this message in the past, but was never able to locate the source. It may be caused by GitGutter, not sure.

At least we know for sure something is interfering with GitGutter and we don't need to bother the core devs any further with that issue. I'll try to locate the source of the message. Maybe it points me to the reason.

**Comment by deathaxe** [comment]
I was just able to reproduce the issue by luck. I have no idea what triggered it, but I was able to localize the issue to the following line:

https://github.com/jisaacks/GitGutter/blob/d9d9fb804f5d61c13ba2f8746af33a9f3c609075/modules/handler.py#L726

`set_timeout()` is an alias for `sublime.set_timeout_async()` in ST3. 

I expanded the function and added some debug prints to find where exactly it gets interrupted.

```python
        print("execute_async:", args)

        def executor(resolve):
            print("  -> executor")

            def on_time():
                print("  -> on_time")
                return resolve(self.execute(args, decode))

            set_timeout(on_time, 10)
            print("  -> set_timeout called!")

        return Promise(executor)
```

If I open the command panel and call "GitGutter: Compare against Commits" the console shows the following lines:

```
GitGutter: Query commits...
execute_async: ['git', 'log', '--all', '--pretty=%h %s\x07%an <%aE>\x07%ad (%ar)', '--date=local', '--max-count=9000']
  -> executor
  -> set_timeout called!
```

This means the issue is located in the `sublime.set_timeout_async()` API function. GitGutter requests it to call `on_time()` a little later, but it waits for the next keypress/mouseclick to do so.

Once a key is pressed, console shows:

```
  -> on_time
```

Maybe we need to ask @wbond what could cause this behavior. 

The little time delay of 10ms should enable the plugin_host to trigger other plugin's functions between several git calls. If I reduce this time to 0 or increase it to 50 or so, `on_time()` is called correctly on my Windows box. 

Not sure whether it does on OSX, too.

**Comment by deathaxe** [comment]
@SCdF: I pushed a commit to set the delay to 0. Could you download the `debug.zip` from the post above again and try, whether the panel appears immediately or not?

**Comment by SCdF** [comment]
@deathaxe this now works for me both extracted out to the Packages directory, and as an Installed Package. Nice job!

**Comment by deathaxe** [comment]
Ok, will add it to the master check some things and publish it as ordinary release soon.

**Comment by wbond** [comment]
A timeout of `0` will effectively run the command at the end of the current unit of work. A delay of 10 used to be necessary to get around some subtle bugs in the API of ST2, but I believe Jon fixed the bug quite a while ago.

I haven't had time to fully understand the context of the snippet of code above, but it does look like an example where you may run into issues with the variables bound to the closure being modified by another bit of code.

**Comment by deathaxe** [comment]
> I haven't had time to fully understand the context of the snippet

Those structures were introduced with the use of the `Promise` class by @rchl a year ago. I first found this concept quite useful to chain asynchronous functions together, but today I find it a bit (too) complicated for that task and not the way python requires it. I honestly didn't catch all aspects of that Promise structures as well. They indeed cause some kind of JavaScript like confusion which is not the python way of life. 😉 😄  Maybe I can find a better, smaller and more pythonic solution for that some day.

> ... variables bound to the closure being modified by another bit of code.

Sounds logical somehow and would answer the question, why this happens when calling the command from command pallet only. Calling them from console or per key binding works properly.

_My idea:_ If e.g. `git_gutter_compare_commit` is called from the command pallet an on_activate() event sent to the view as soon as the command pallet closes. This causes the EventListener to call the main `git_gutter` command to update the gutter icons and so on. At the nearly same time, the `git_gutter_compare_commit` runs `git_gutter`, too. Both end up in running some git commands via the `GitGutterHandler` class using the snippet from above.

So with the little delay both fired `git_gutter` commands may confuse the `set_timeout` function.

But there keeps one question open with it: If I explicitly use `sublime.set_timeout()` it works well even with 10ms delay. The issue is caused by `sublime.set_timeout_async()` only. Does that happen by luck or is it due to some kind of inter-process-communication or so?

Hmm! With this situation and its side effects in mind I might end up in abandoning the Promise stuff and use a dedicated threads and queues to run everything after each other in a determined way without pushing anything to the `set_timeout` function.
